torch/_inductor/scheduler.py

```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
 
         return sorted(possible_fusions, key=self.score_fusion_key, reverse=True)
 
+    @staticmethod
+    def _upper_triangle_pairs(nodes):
+        for node1_index, node1 in enumerate(nodes):
+            for node2 in nodes[node1_index + 1 :]:
+                yield node1, node2
+
+    @staticmethod
+    def reorder_deps(dep1, dep2) -> Optional[List[int]]:
+        if dep1 == dep2:
+            return None
+
+        if not isinstance(dep1, MemoryDep) or not isinstance(dep2, MemoryDep):
+            return None
+
+        sizevars = V.graph.sizevars
+
+        def expr_key(expr):
+            return (sizevars.size_hint(expr), str(expr))
+
+        def sort_exprs(exprs):
+            return sorted(exprs, key=expr_key)
+
+        if sort_exprs(dep1.size) != sort_exprs(dep2.size):
+            return None
+
+        ndim = len(dep1.size)
+        c = canonicalization_prefix()
+        index_vars = [sympy_symbol(f"{c}{i}") for i in range(ndim)]
+        strides1 = sizevars.stride_vars(dep1.index, index_vars)
+        strides2 = sizevars.stride_vars(dep2.index, index_vars)
+
+        dep1_permutation = sorted(range(ndim), key=lambda i: expr_key(strides1[i]))
+        dep2_permutation = sorted(range(ndim), key=lambda i: expr_key(strides2[i]))
+        if dep1_permutation == dep2_permutation:
+            return None
+
+        valid = all(
+            dep1.size[d1] == dep2.size[d2]
+            for d1, d2 in zip(dep1_permutation, dep2_permutation)
+        )
+        if not valid:
+            return None
+
+        # Sanity check, if we apply the reordering to the index expression, it should be equal
+        dep2_reindexed = sympy_subs(
+            dep2.index,
+            {
+                index_vars[d2]: index_vars[d1]
+                for d1, d2 in zip(dep1_permutation, dep2_permutation)
+            },
+        )
+        if dep1.index != dep2_reindexed:
+            return None
```

Comment:
Should we make this an assert?

Reply:
`stride_vars` can give the wrong answer in some edge cases (which I documented in #90184), so this check can fail for that reason.

Reply:
oof. Perhaps leave a comment?